File: laocai-bot-spring-boot-starter/src/main/java/indi/dkx/laocai/bot/model/segment/FaceSegment.java
package indi.dkx.laocai.bot.model.segment;

import com.fasterxml.jackson.annotation.JsonIgnore;

import java.util.Objects;

/**
 * 表情消息段。
 */
public final class FaceSegment implements Segment {

    record Data(String faceId, boolean isLarge) {
        FaceSegment toSegment() {
            return new FaceSegment(this);
        }
    }

    private final Data data;

    private FaceSegment(Data data) {
        this.data = data;
    }

    public static FaceSegment of(String faceId, boolean isLarge) {
        return new FaceSegment(new Data(Objects.requireNonNull(faceId), isLarge));
    }

    @Override
    public String getType() {
        return "face";
    }

    @Override
    public Data getData() {
        return data;
    }

    @JsonIgnore
    public String getFaceId() {
        return data.faceId();
    }

    @JsonIgnore
    public boolean isLarge() {
        return data.isLarge();
    }
}

File: laocai-bot-spring-boot-starter/src/main/kotlin/indi/kyson/laocai/bot/segment/FaceSegment.kt
package indi.kyson.laocai.bot.segment

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty

/**
 * 表情消息段。
 */
class FaceSegment private constructor(override val data: Data) : Segment {

    data class Data(
        @JsonProperty("face_id") val faceId: String,
        @JsonProperty("is_large") val isLarge: Boolean,
    ) {
        fun toSegment(): FaceSegment = FaceSegment(this)
    }

    override val type: String
        get() = "face"

    @get:JsonIgnore
    val faceId: String
        get() = data.faceId

    @get:JsonIgnore
    val isLarge: Boolean
        get() = data.isLarge

    companion object {
        @JvmStatic
        fun of(faceId: String, isLarge: Boolean): FaceSegment = FaceSegment(Data(faceId, isLarge))
    }
}
